README.md
[![Build & Test][github-actions-badge]][github-actions-link]
[![crates.io][crate-badge]][crate-link]
[![Docs][docs-badge]][docs-link]

# libmudtelnet

A low-level Telnet protocol implementation for MUD clients, written in Rust.

`libmudtelnet` is a fork of [libtelnet-rs], which is itself inspired by the [libtelnet].

# Changelog

See [CHANGELOG.md](CHANGELOG.md).

# Usage

Check [src/tests.rs](tests/tests.rs) for an example parser. For a larger
example, see the [Blightmud] MUD client that uses `libtelnet-rs` for its Telnet
handling.

First, construct a parser with [`Parser::new()`][new-parser]. Ideally, you would
place this parser somewhere directly behind a socket or external source of data.

When data comes in from the socket, immediately send it into the parser with
[`parser.receive(data)`][receive]. This will append it to the current internal
buffer, and process and return any [telnet events] to be looped over and handled
as your application requires.

Any text to be sent back over the socket to the remote end should be sent
through [`parser.send_text(data)`][send-text] to ensure data will be encoded
properly for the telnet protocol. Data to be sent will be provided either by
a `events::TelnetEvents::DataSend` event after processing, or as a return from
any method used for sending data.

# Compatibility

The initial release of `libmudtelnet` has been tested for compatibility with
`libtelnet-rs`. In general while much of the code has been rewritten to be more
idiomatic Rust, the API is the same and breaking changes have been avoided. An
upcoming semver incompatible release will be made with broader API changes in
the near future.

See [CHANGELOG.md](CHANGELOG.md) for more details.

# Credits

Many thanks to:

* [envis10n] for his work on [libtelnet-rs], which `libmudtelnet` is forked from.
* [Sean Middleditch] for his work on [libtelnet], which inspired `libtelnet-rs`.

[github-actions-badge]: https://img.shields.io/github/actions/workflow/status/blightmud/libmudtelnet/rust.yml?label=build%20%26%20tests&logo=github&style=for-the-badge&branch=main
[github-actions-link]: https://github.com/blightmud/libmudtelnet/actions/workflows/rust.yml?query=branch%3Amain
[crate-badge]: https://img.shields.io/crates/v/libmudtelnet.svg?style=for-the-badge
[crate-link]: https://crates.io/crates/libmudtelnet
[docs-badge]: https://img.shields.io/badge/docs-rs-blue.svg?style=for-the-badge
[docs-link]: https://docs.rs/libmudtelnet
[libtelnet-rs]: https://github.com/envis10n/libtelnet-rs
[libtelnet]: https://github.com/seanmiddleditch/libtelne
[Blightmud]: https://github.com/blightmud/blightmud
[new-parser]: https://docs.rs/libmudtelnet/latest/libmudtelnet/struct.Parser.html#method.new
[receive]: https://docs.rs/libmudtelnet/latest/libmudtelnet/struct.Parser.html#method.receive
[telnet events]: https://docs.rs/libmudtelnet/latest/libmudtelnet/events/enum.TelnetEvents.html
[send-text]: https://docs.rs/libmudtelnet/latest/libmudtelnet/struct.Parser.html#method.send_text
[envis10n]: https://github.com/envis10n/
[Sean Middleditch]: https://github.com/seanmiddleditch/


CHANGELOG.md
# v2.0.1 (pending)

Initial release of `libmudtelnet` - a fork of `libtelnet-rs`.

## Bug fixes

* A `SE` byte that isn't preceded by `IAC` is now properly handled as a normal
  byte during Telnet subnegotiation processing.
* Fixed a panic when Telnet option code 0xFF is negotiated, and a truncated 
  subnegotiation (e.g. `IAC SB IAC SE`) is received.
* Multiple escaped `IAC` bytes (`e.g. IAC IAC IAC IAC`) are now properly 
 unescaped (e.g. `IAC IAC`) instead of truncated (e.g. `IAC`). 

## Features

* Many API types now derive helpful traits (`Debug`, `Eq`, etc.).
* An optional `arbitrary` crate feature is now available to enable generating 
  arbitrary `event::TelnetIAC`, `event::TelnetNegotiation` and `event::TelnetSubnegotiation` 
  instances for testing.

## Misc

* CI improvements.
* Fuzz testing.
* Simple benchmarking.
* Small performance and safety improvements (avoiding direct indexing, etc.).

# v2.0.0

As of version 2.0, `libtelnet_rs` has switched over to using the `bytes`
[crate](https://crates.io/crates/bytes).

With this change, the method signatures for most methods that return `Vec<u8>`
now return `Bytes` instead.

For most situations where a `Vec<u8>` would be required, the returned value can
simply be changed to utilize `Bytes::to_vec()`.

To make usage of the new dependency a little bit easier, it is also re-exported
as `libtelnet_rs::bytes`.


tests/tests.rs
use bytes::Bytes;

use libmudtelnet::compatibility::{CompatibilityEntry, CompatibilityTable};
use libmudtelnet::events::{TelnetEvents, TelnetSubnegotiation};
use libmudtelnet::telnet::{op_command as cmd, op_option as opt};
use libmudtelnet::Parser;

/// Test the parser and its general functionality.

#[derive(PartialEq, Debug)]
enum Event {
  Iac,
  Negotiation,
  Subnegotiation,
  Recv,
  Send,
  Decom,
}

macro_rules! events {
    ( $( $x:expr ),* ) => {
        {
            #[allow(unused_mut)]
            let mut temp_ce = CapturedEvents::default();
            $(
                temp_ce.push($x);
            )*
            temp_ce
        }
    };
}

#[derive(Default, Debug)]
struct CapturedEvents {
  events: Vec<Event>,
}

impl CapturedEvents {
  fn push(&mut self, event: Event) {
    self.events.push(event);
  }
}

impl PartialEq for CapturedEvents {
  fn eq(&self, other: &Self) -> bool {
    if self.events.len() == other.events.len() {
      self
        .events
        .iter()
        .zip(other.events.iter())
        .all(|(val1, val2)| val1 == val2)
    } else {
      false
    }
  }
}

fn handle_events(event_list: Vec<TelnetEvents>) -> CapturedEvents {
  let mut events = CapturedEvents::default();
  for event in event_list {
    match event {
      TelnetEvents::IAC(ev) => {
        println!("IAC: {}", ev.command);
        events.push(Event::Iac);
      }
      TelnetEvents::Negotiation(ev) => {
        println!("Negotiation: {} {}", ev.command, ev.option);
        events.push(Event::Negotiation);
      }
      TelnetEvents::Subnegotiation(ev) => {
        println!("Subnegotiation: {} {:?}", ev.option, ev.buffer);
        events.push(Event::Subnegotiation);
      }
      TelnetEvents::DataReceive(buffer) => {
        println!(
          "Receive: {}",
          std::str::from_utf8(&buffer[..]).unwrap_or("Bad utf-8 bytes")
        );
        events.push(Event::Recv);
      }
      TelnetEvents::DataSend(buffer) => {
        println!("Send: {:?}", buffer);
        events.push(Event::Send);
      }
      TelnetEvents::DecompressImmediate(buffer) => {
        println!("DECOMPRESS: {:?}", buffer);
        events.push(Event::Decom);
      }
    };
  }
  events
}

#[test]
fn test_parser() {
  let mut instance: Parser = Parser::new();
  instance.options.support_local(opt::GMCP);
  instance.options.support_local(opt::MCCP2);
  if let Some(ev) = instance._will(opt::GMCP) {
    assert_eq!(handle_events(vec![ev]), events![Event::Send]);
  }
  if let Some(ev) = instance._will(opt::MCCP2) {
    assert_eq!(handle_events(vec![ev]), events![Event::Send]);
  }
  assert_eq!(
    handle_events(instance.receive(&[b"Hello, rust!", &[cmd::IAC, cmd::GA][..]].concat())),
    events![Event::Recv, Event::Iac]
  );
  assert_eq!(
    handle_events(instance.receive(&[cmd::IAC, cmd::DO, opt::GMCP])),
    events![]
  );
  assert_eq!(
    handle_events(instance.receive(&[&[cmd::IAC, cmd::DO, 200][..], b"Some random data"].concat())),
    events![Event::Send, Event::Recv]
  );
  assert_eq!(
    handle_events(instance.receive(
      &TelnetSubnegotiation::new(opt::GMCP, Bytes::copy_from_slice(b"Core.Hello {}")).to_bytes()
    ),),
    events![Event::Subnegotiation]
  );
  assert_eq!(
    handle_events(
      instance.receive(
        &[
          &TelnetSubnegotiation::new(opt::GMCP, Bytes::copy_from_slice(b"Core.Hello {}"))
            .to_bytes()[..],
          b"Random text",
          &[cmd::IAC, cmd::GA][..]
        ]
        .concat()
      ),
    ),
    events![Event::Subnegotiation, Event::Recv, Event::Iac]
  );
  assert_eq!(
    handle_events(
      instance.receive(
        &[
          &TelnetSubnegotiation::new(opt::MCCP2, Bytes::copy_from_slice(b" ")).to_bytes()[..],
          b"This is compressed data",
          &[cmd::IAC, cmd::GA][..]
        ]
        .concat()
      ),
    ),
    events![Event::Subnegotiation, Event::Decom]
  );
  assert_eq!(
    // TODO(@cpu): Can data be made easier to understand at a glance?
    handle_events(instance.receive(&[
      87, 104, 97, 116, 32, 105, 115, 32, 121, 111, 117, 114, 32, 112, 97, 115, 115, 119, 111, 114,
      100, 63, 32, 255, 239, 255, 251, 1
    ])),
    events![Event::Recv, Event::Iac, Event::Send]
  );
}

#[test]
fn test_subneg_separate_receives() {
  let mut instance: Parser = Parser::with_capacity(10);
  instance.options.support_local(opt::GMCP);
  instance._will(opt::GMCP);
  let mut events = instance.receive(
    &[
      &[cmd::IAC, cmd::SB, opt::GMCP][..],
      b"Otion.Data { some: json, data: in, here: ! }",
    ]
    .concat(),
  );
  assert_eq!(handle_events(events), events![]);

  events = instance.receive(b"More.Data { some: json, data: in, here: ! }");
  assert_eq!(handle_events(events), events![]);

  events = instance.receive(
    &[
      &[cmd::IAC, cmd::SE][..],
      &[cmd::IAC, cmd::SB, opt::GMCP][..],
      b"Otion.Data { some: json, data: in, here: ! }",
    ]
    .concat(),
  );
  assert_eq!(handle_events(events), events![Event::Subnegotiation]);

  events = instance.receive(
    &[
      b"More.Data { some: json, data: in, here: ! }",
      &[cmd::IAC, cmd::SE][..],
    ]
    .concat(),
  );
  assert_eq!(handle_events(events), events![Event::Subnegotiation]);
}

// Test that receiving a subnegotiation with embedded UTF-8 content works correctly,
// even when the content includes a SE byte.
#[test]
fn test_subneg_utf8_content() {
  use cmd::{IAC, SB, SE};
  use opt::GMCP;

  // Create a parser that will support GMCP.
  let mut parser = Parser::new();
  parser.options.support_local(GMCP);
  parser._will(GMCP);

  // Construct a GMCP message containing a UTF-8 sequence that happens
  // to include SE (0xF0). This should be permitted as long as the SE isn't
  // preceeded by IAC (0xFF). For our test case we'll use the content
  // '👋' (0xF0, 0x9F, 0x91, 0x8B) - where the leading byte is SE.
  let prefix = &[IAC, SB, GMCP][..];
  let wave_emoji = &[0xF0, 0x9F, 0x91, 0x8B][..];
  let suffix = &[IAC, SE][..];
  let gmcp_msg = [prefix, wave_emoji, suffix].concat();

  // Receive the GMCP message with the parser. This should produce one event.
  let events = parser.receive(&gmcp_msg);
  assert_eq!(events.len(), 1, "only expected one event to be parsed");

  // The event should be a Subnegotiation for the GMCP option, with the correct in-tact
  // buffer contents.
  if let TelnetEvents::Subnegotiation(sub) = events.first().unwrap() {
    assert_eq!(sub.option, 201, "option should be GMCP");
    assert_eq!(
      sub.buffer, wave_emoji,
      "buffer should be equal to the wave emoji"
    );
  } else {
    panic!("missing expected DataReceive event");
  }
}

/// Test escaping IAC bytes in a buffer.
#[test]
fn test_escape() {
  let initial = vec![
    cmd::IAC,
    cmd::SB,
    201,
    cmd::IAC,
    205,
    202,
    cmd::IAC,
    cmd::SE,
  ];
  let expected = Bytes::copy_from_slice(&[
    cmd::IAC,
    cmd::IAC,
    cmd::SB,
    201,
    cmd::IAC,
    cmd::IAC,
    205,
    202,
    cmd::IAC,
    cmd::IAC,
    cmd::SE,
  ]);
  assert_eq!(expected, Parser::escape_iac(initial.clone()));
  assert_eq!(initial, Parser::unescape_iac(expected));
}

/// Test unescaping IAC bytes in a buffer.
#[test]
fn test_unescape() {
  let initial = vec![
    cmd::IAC,
    cmd::IAC,
    cmd::SB,
    201,
    cmd::IAC,
    cmd::IAC,
    205,
    202,
    cmd::IAC,
    cmd::IAC,
    cmd::SE,
  ];
  let expected = Bytes::copy_from_slice(&[
    cmd::IAC,
    cmd::SB,
    201,
    cmd::IAC,
    205,
    202,
    cmd::IAC,
    cmd::SE,
  ]);
  assert_eq!(expected, Parser::unescape_iac(initial.clone()));
  assert_eq!(initial, Parser::escape_iac(expected));
}

#[test]
fn test_escape_roundtrip_bug_one() {
  // The original libtelnet-rs mishandles this input:
  let data = vec![cmd::IAC, cmd::IAC, 228];
  let escaped = Parser::escape_iac(data.clone());
  let unescaped = Parser::unescape_iac(escaped);
  assert_eq!(unescaped, data);
}

#[test]
fn test_escape_roundtrip_bug_two() {
  // The original libtelnet-rs mishandles this input:
  let data = vec![228, cmd::IAC, cmd::IAC];
  let escaped = Parser::escape_iac(data.clone());
  let unescaped = Parser::unescape_iac(escaped);
  assert_eq!(unescaped, data);
}

#[test]
fn test_bad_subneg_dbuffer() {
  // Configure opt 0xFF (IAC) as local supported, and local state enabled.
  let entry = CompatibilityEntry::new(true, false, true, false);
  let opts = CompatibilityTable::from_options(&[(cmd::IAC, entry.into_u8())]);
  // Receive a malformed subnegotiation - this should not panic.
  Parser::with_support(opts).receive(&[cmd::IAC, cmd::SB, cmd::IAC, cmd::SE]);
}

#[test]
fn test_into_bytes() {
  let bytes = libmudtelnet::events::TelnetIAC::new(cmd::IAC).to_bytes();
  assert!(!bytes.is_empty())
}

#[cfg(test)]
mod compat_tests {
  use compat::{test_app, TelnetApplication};

  #[test]
  fn test_parser_diff1() {
    test_app(&TelnetApplication {
      options: vec![(255, 254)],
      received_data: vec![vec![255, 255, 255, 255, 255, 254, 255, 0]],
    });
  }

  #[test]
  fn test_parser_diff2() {
    test_app(&TelnetApplication {
      options: vec![],
      received_data: vec![vec![45, 255, 250, 255]],
    });
  }

  #[test]
  fn test_parser_diff3() {
    test_app(&TelnetApplication {
      options: vec![(0, 1)],
      received_data: vec![vec![255, 253, 0]],
    })
  }

  #[test]
  fn test_parser_diff4() {
    test_app(&TelnetApplication {
      options: vec![],
      received_data: vec![vec![255, 250, 255, 255, 240, 250]],
    })
  }

  #[test]
  fn test_parser_diff5() {
    test_app(&TelnetApplication {
      options: vec![],
      received_data: vec![vec![255, 250, 255, 240, 0]],
    })
  }

  #[test]
  fn test_parser_diff6() {
    test_app(&TelnetApplication {
      options: vec![],
      received_data: vec![vec![240, 255, 250, 255, 240, 0]],
    })
  }

  #[test]
  fn test_parser_diff7() {
    test_app(&TelnetApplication {
      options: vec![],
      received_data: vec![vec![255]],
    })
  }

  #[test]
  fn test_parser_diff8() {
    test_app(&TelnetApplication {
      options: vec![],
      received_data: vec![vec![255, 252, 0]],
    })
  }

  #[test]
  fn test_parser_diff9() {
    test_app(&TelnetApplication {
      options: vec![],
      received_data: vec![vec![254, 255, 255, 255, 254, 0]],
    })
  }

  #[test]
  fn test_parser_diff10() {
    test_app(&TelnetApplication {
      options: vec![(255, 254), (1, 0)],
      received_data: vec![vec![255, 253, 255]],
    })
  }
}


src/events.rs
use alloc::vec::Vec;

use bytes::{BufMut, Bytes, BytesMut};

use crate::telnet::op_command::{IAC, SB, SE};
use crate::Parser;

/// A struct representing a 2 byte IAC sequence.
#[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
#[cfg_attr(feature = "arbitrary", derive(arbitrary::Arbitrary))]
pub struct TelnetIAC {
  pub command: u8,
}

impl TelnetIAC {
  #[must_use]
  pub fn new(command: u8) -> Self {
    Self { command }
  }

  /// Consume the sequence struct and return the bytes.
  #[must_use]
  pub fn to_bytes(self) -> Bytes {
    Bytes::copy_from_slice(&[IAC, self.command])
  }

  #[must_use]
  #[deprecated(since = "0.2.1", note = "Use `to_bytes` instead.")]
  pub fn into_bytes(self) -> Bytes {
    self.to_bytes()
  }
}

/// A struct representing a 3 byte IAC sequence.
#[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
#[cfg_attr(feature = "arbitrary", derive(arbitrary::Arbitrary))]
pub struct TelnetNegotiation {
  pub command: u8,
  pub option: u8,
}

impl TelnetNegotiation {
  #[must_use]
  pub fn new(command: u8, option: u8) -> Self {
    Self { command, option }
  }

  /// Consume the sequence struct and return the bytes.
  #[must_use]
  pub fn to_bytes(self) -> Bytes {
    Bytes::copy_from_slice(&[IAC, self.command, self.option])
  }

  #[must_use]
  #[deprecated(since = "0.2.1", note = "Use `to_bytes` instead.")]
  pub fn into_bytes(self) -> Bytes {
    self.to_bytes()
  }
}

/// A struct representing an arbitrary length IAC subnegotiation sequence.
#[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
pub struct TelnetSubnegotiation {
  pub option: u8,
  pub buffer: Bytes,
}

#[cfg(feature = "arbitrary")]
impl<'a> arbitrary::Arbitrary<'a> for TelnetSubnegotiation {
  fn arbitrary(u: &mut arbitrary::Unstructured<'a>) -> arbitrary::Result<Self> {
    let option = u.arbitrary()?;
    let buffer: Vec<u8> = u.arbitrary()?;
    Ok(Self {
      option,
      buffer: Bytes::from(buffer),
    })
  }
}

impl TelnetSubnegotiation {
  pub fn new(option: u8, buffer: Bytes) -> Self {
    Self { option, buffer }
  }

  #[must_use]
  pub fn to_bytes(self) -> Bytes {
    let head = [IAC, SB, self.option];
    let parsed = &Parser::escape_iac(self.buffer)[..];
    let tail = [IAC, SE];
    let mut buf = BytesMut::with_capacity(head.len() + parsed.len() + tail.len());
    buf.put(&head[..]);
    buf.put(parsed);
    buf.put(&tail[..]);
    buf.freeze()
  }

  #[must_use]
  #[deprecated(since = "0.2.1", note = "Use `to_bytes` instead.")]
  pub fn into_bytes(self) -> Bytes {
    self.to_bytes()
  }
}

/// An enum representing various telnet events.
#[derive(Clone, Debug, Eq, PartialEq)]
pub enum TelnetEvents {
  /// An IAC command sequence.
  IAC(TelnetIAC),
  /// An IAC negotiation sequence.
  Negotiation(TelnetNegotiation),
  /// An IAC subnegotiation sequence.
  Subnegotiation(TelnetSubnegotiation),
  /// Regular data received from the remote end.
  DataReceive(Bytes),
  /// Any data to be sent to the remote end.
  DataSend(Bytes),
  /// MCCP2/3 compatibility. MUST DECOMPRESS THIS DATA BEFORE PARSING
  DecompressImmediate(Bytes),
}

impl From<TelnetIAC> for TelnetEvents {
  fn from(iac: TelnetIAC) -> Self {
    TelnetEvents::IAC(iac)
  }
}

impl From<TelnetNegotiation> for TelnetEvents {
  fn from(neg: TelnetNegotiation) -> Self {
    TelnetEvents::Negotiation(neg)
  }
}

impl From<TelnetSubnegotiation> for TelnetEvents {
  fn from(sub: TelnetSubnegotiation) -> Self {
    TelnetEvents::Subnegotiation(sub)
  }
}

impl TelnetEvents {
  /// Helper method to generate a `TelnetEvents::DataSend`.
  #[deprecated(since = "0.2.1", note = "Construct enum variant directly or use into.")]
  pub fn build_send(buffer: Bytes) -> Self {
    TelnetEvents::DataSend(buffer)
  }

  /// Helper method to generate a `TelnetEvents::DataReceive`.
  #[deprecated(since = "0.2.1", note = "Construct enum variant directly or use into.")]
  pub fn build_receive(buffer: Bytes) -> Self {
    TelnetEvents::DataReceive(buffer)
  }

  /// Helper method to generate a `TelnetEvents::IAC`.
  #[must_use]
  #[deprecated(since = "0.2.1", note = "Construct enum variant directly or use into.")]
  pub fn build_iac(command: u8) -> TelnetEvents {
    TelnetEvents::IAC(TelnetIAC::new(command))
  }

  /// Helper method to generate a `TelnetEvents::Negotiation`.
  #[must_use]
  #[deprecated(since = "0.2.1", note = "Construct enum variant directly or use into.")]
  pub fn build_negotiation(command: u8, option: u8) -> Self {
    TelnetEvents::Negotiation(TelnetNegotiation::new(command, option))
  }

  /// Helper method to generate a `TelnetEvents::Subnegotiation`.
  #[deprecated(since = "0.2.1", note = "Construct enum variant directly or use into.")]
  pub fn build_subnegotiation(option: u8, buffer: Bytes) -> Self {
    TelnetEvents::Subnegotiation(TelnetSubnegotiation::new(option, buffer))
  }

  #[must_use]
  pub fn to_bytes(self) -> Bytes {
    match self {
      TelnetEvents::IAC(iac) => iac.to_bytes(),
      TelnetEvents::Negotiation(neg) => neg.to_bytes(),
      TelnetEvents::Subnegotiation(sub) => sub.to_bytes(),
      TelnetEvents::DataReceive(data)
      | TelnetEvents::DataSend(data)
      | TelnetEvents::DecompressImmediate(data) => data,
    }
  }
}

/*
TODO(@cpu): remove/retool this stuff in breaking release.
*/
#[allow(clippy::from_over_into)]
impl Into<Bytes> for TelnetIAC {
  fn into(self) -> Bytes {
    self.to_bytes()
  }
}

#[allow(clippy::from_over_into)]
impl Into<Vec<u8>> for TelnetIAC {
  fn into(self) -> Vec<u8> {
    self.to_bytes().into()
  }
}

#[allow(clippy::from_over_into)]
impl Into<Bytes> for TelnetNegotiation {
  fn into(self) -> Bytes {
    self.to_bytes()
  }
}

#[allow(clippy::from_over_into)]
impl Into<Vec<u8>> for TelnetNegotiation {
  fn into(self) -> Vec<u8> {
    self.to_bytes().into()
  }
}

#[allow(clippy::from_over_into)]
impl Into<Bytes> for TelnetSubnegotiation {
  fn into(self) -> Bytes {
    self.to_bytes()
  }
}

#[allow(clippy::from_over_into)]
impl Into<Vec<u8>> for TelnetSubnegotiation {
  fn into(self) -> Vec<u8> {
    self.to_bytes().into()
  }
}


src/compatibility.rs
/// An expansion of a bitmask contained in `CompatibilityTable`.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct CompatibilityEntry {
  /// Whether we support this option from us -> them.
  pub local: bool,
  /// Whether we support this option from them -> us.
  pub remote: bool,
  /// Whether this option is locally enabled.
  pub local_state: bool,
  /// Whether this option is remotely enabled.
  pub remote_state: bool,
}

impl CompatibilityEntry {
  #[must_use]
  pub fn new(local: bool, remote: bool, local_state: bool, remote_state: bool) -> Self {
    Self {
      local,
      remote,
      local_state,
      remote_state,
    }
  }

  /// Creates a u8 bitmask from this entry.
  #[must_use]
  pub fn into_u8(self) -> u8 {
    let mut res = 0;
    if self.local {
      res |= CompatibilityTable::ENABLED_LOCAL;
    }
    if self.remote {
      res |= CompatibilityTable::ENABLED_REMOTE;
    }
    if self.local_state {
      res |= CompatibilityTable::LOCAL_STATE;
    }
    if self.remote_state {
      res |= CompatibilityTable::REMOTE_STATE;
    }
    res
  }

  /// Expands a u8 bitmask into a `CompatibilityEntry`.
  #[must_use]
  pub fn from(value: u8) -> Self {
    Self {
      local: value & CompatibilityTable::ENABLED_LOCAL == CompatibilityTable::ENABLED_LOCAL,
      remote: value & CompatibilityTable::ENABLED_REMOTE == CompatibilityTable::ENABLED_REMOTE,
      local_state: value & CompatibilityTable::LOCAL_STATE == CompatibilityTable::LOCAL_STATE,
      remote_state: value & CompatibilityTable::REMOTE_STATE == CompatibilityTable::REMOTE_STATE,
    }
  }
}

/// A table of options that are supported locally or remotely, and their current state.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct CompatibilityTable {
  options: [u8; 256],
}

impl Default for CompatibilityTable {
  fn default() -> Self {
    Self { options: [0; 256] }
  }
}

impl CompatibilityTable {
  /// Option is locally supported.
  pub const ENABLED_LOCAL: u8 = 1;
  /// Option is remotely supported.
  pub const ENABLED_REMOTE: u8 = 1 << 1;
  /// Option is currently enabled locally.
  pub const LOCAL_STATE: u8 = 1 << 2;
  /// Option is currently enabled remotely.
  pub const REMOTE_STATE: u8 = 1 << 3;

  #[must_use]
  pub fn new() -> Self {
    Self::default()
  }

  /// Create a table with some option values set.
  ///
  /// # Arguments
  ///
  /// `values` - A slice of `(u8, u8)` tuples. The first value is the option code, and the second is the bitmask value for that option.
  ///
  /// # Notes
  ///
  /// An option bitmask can be generated using the `CompatibilityEntry` struct, using `entry.into_u8()`.
  #[must_use]
  pub fn from_options(values: &[(u8, u8)]) -> Self {
    let mut options = [0; 256];
    for (opt, val) in values {
      options[*opt as usize] = *val;
    }
    Self { options }
  }

  /// Enable local support for an option.
  pub fn support_local(&mut self, option: u8) {
    let mut opt = CompatibilityEntry::from(self.options[option as usize]);
    opt.local = true;
    self.set_option(option, opt);
  }

  /// Enable remote support for an option.
  pub fn support_remote(&mut self, option: u8) {
    let mut opt = CompatibilityEntry::from(self.options[option as usize]);
    opt.remote = true;
    self.set_option(option, opt);
  }

  /// Enable both remote and local support for an option.
  pub fn support(&mut self, option: u8) {
    let mut opt = CompatibilityEntry::from(self.options[option as usize]);
    opt.local = true;
    opt.remote = true;
    self.set_option(option, opt);
  }

  /// Retrieve a `CompatbilityEntry` generated from the current state of the option value.
  #[must_use]
  pub fn get_option(&self, option: u8) -> CompatibilityEntry {
    CompatibilityEntry::from(self.options[option as usize])
  }

  /// Set an option value by getting the bitmask from a `CompatibilityEntry`.
  pub fn set_option(&mut self, option: u8, entry: CompatibilityEntry) {
    self.options[option as usize] = entry.into_u8();
  }

  /// Reset all negotiated states
  pub fn reset_states(&mut self) {
    for opt in &mut self.options {
      let mut entry = CompatibilityEntry::from(*opt);
      entry.local_state = false;
      entry.remote_state = false;
      *opt = entry.into_u8();
    }
  }
}

#[cfg(test)]
mod test_compat {
  use super::*;
  use crate::telnet::op_option::GMCP;

  #[test]
  fn test_reset() {
    let mut table = CompatibilityTable::default();
    let entry = CompatibilityEntry::new(true, true, true, true);
    assert!(entry.remote);
    assert!(entry.local);
    assert!(entry.remote_state);
    assert!(entry.local_state);
    table.set_option(GMCP, entry);
    table.reset_states();
    let entry = table.get_option(GMCP);
    assert!(entry.remote);
    assert!(entry.local);
    assert!(!entry.remote_state);
    assert!(!entry.local_state);
  }
}


compat/benches/benchmark.rs
use arbitrary::{Arbitrary, Unstructured};
use bencher::{benchmark_group, benchmark_main, Bencher};
use libmudtelnet::bytes::Bytes;
use libmudtelnet::compatibility::CompatibilityEntry;
use libmudtelnet::events::{TelnetIAC, TelnetNegotiation, TelnetSubnegotiation};
use rand::rngs::StdRng;
use rand::{RngCore, SeedableRng};
use std::hint::black_box;

#[derive(Debug, Arbitrary)]
enum SimulationEvents {
  IAC(TelnetIAC),
  Negotiation(TelnetNegotiation),
  Subnegotiation(TelnetSubnegotiation),
  Data(Vec<u8>),
}

impl SimulationEvents {
  fn to_bytes(&self) -> Bytes {
    // TODO(XXX): to_bytes() shouldn't require moving, we could drop clones here. Or maybe there
    //   should be an as_ref() for each event?
    match self {
      SimulationEvents::IAC(iac) => iac.clone().to_bytes(),
      SimulationEvents::Negotiation(neg) => neg.clone().to_bytes(),
      SimulationEvents::Subnegotiation(sub) => sub.clone().to_bytes(),
      SimulationEvents::Data(data) => Bytes::copy_from_slice(data),
    }
  }
}

const SEED: [u8; 32] = [
  0xb8, 0x5a, 0x8d, 0xcf, 0xb9, 0x9f, 0x1a, 0x2a, 0x6, 0xf8, 0x46, 0x14, 0xee, 0x5f, 0x69, 0xae,
  0x44, 0x23, 0x36, 0xf9, 0xc0, 0x40, 0x57, 0xd8, 0x62, 0x7f, 0x98, 0x71, 0xe2, 0xba, 0x76, 0xf9,
];

fn bench_receive(b: &mut Bencher) {
  let mut data_buf = [0; 1024 * 1024];
  StdRng::from_seed(SEED).fill_bytes(&mut data_buf);

  b.iter(|| {
    let mut parser = libmudtelnet::Parser::default();
    for i in 0..255 {
      parser.options.set_option(
        i,
        CompatibilityEntry {
          local: true,
          remote: true,
          local_state: false,
          remote_state: false,
        },
      );
    }

    const INPUT_EVENTS: usize = 10_000;
    for _ in 0..INPUT_EVENTS {
      let input_event = SimulationEvents::arbitrary(&mut Unstructured::new(&data_buf)).unwrap();
      black_box(parser.receive(&input_event.to_bytes()));
    }
  });
}

fn bench_og_receive(b: &mut Bencher) {
  let mut data_buf = [0; 1024 * 1024];
  StdRng::from_seed(SEED).fill_bytes(&mut data_buf);

  b.iter(|| {
    let mut parser = libtelnet_rs::Parser::default();
    for i in 0..255 {
      parser.options.set_option(
        i,
        libtelnet_rs::compatibility::CompatibilityEntry {
          local: true,
          remote: true,
          local_state: false,
          remote_state: false,
        },
      );
    }

    const INPUT_EVENTS: usize = 10_000;
    for _ in 0..INPUT_EVENTS {
      let input_event = SimulationEvents::arbitrary(&mut Unstructured::new(&data_buf)).unwrap();
      black_box(parser.receive(&input_event.to_bytes()));
    }
  });
}

benchmark_group!(parser_benches, bench_receive, bench_og_receive);

benchmark_main!(parser_benches);


src/telnet.rs
// Define a public u8 constant with the given name and constant value.
macro_rules! u8_const {
  ($name: ident, $value: expr) => {
    pub const $name: u8 = $value;
  };
}

/// Module containing constants for Telnet Command codes.
// TODO(XXX): rename to cmd.
pub mod op_command {
  u8_const!(IAC, 255);
  u8_const!(WILL, 251);
  u8_const!(WONT, 252);
  u8_const!(DO, 253);
  u8_const!(DONT, 254);
  u8_const!(NOP, 241);
  u8_const!(SB, 250);
  u8_const!(SE, 240);
  u8_const!(IS, 0);
  u8_const!(SEND, 1);
  u8_const!(GA, 249);
  u8_const!(EOR, 239);
}

/// Module containing constants for Telnet Option codes.
// TODO(XXX): rename to opt.
pub mod op_option {
  u8_const!(BINARY, 0);
  u8_const!(ECHO, 1);
  u8_const!(RCP, 2);
  u8_const!(SGA, 3);
  u8_const!(NAMS, 4);
  u8_const!(STATUS, 5);
  u8_const!(TM, 6);
  u8_const!(RCTE, 7);
  u8_const!(NAOL, 8);
  u8_const!(NAOP, 9);
  u8_const!(NAOCRD, 10);
  u8_const!(NAOHTS, 11);
  u8_const!(NAOHTD, 12);
  u8_const!(NAOFFD, 13);
  u8_const!(NAOVTS, 14);
  u8_const!(NAOVTD, 15);
  u8_const!(NAOLFD, 16);
  u8_const!(XASCII, 17);
  u8_const!(LOGOUT, 18);
  u8_const!(BM, 19);
  u8_const!(DET, 20);
  u8_const!(SUPDUP, 21);
  u8_const!(SUPDUPOUTPUT, 22);
  u8_const!(SNDLOC, 23);
  u8_const!(TTYPE, 24);
  u8_const!(EOR, 25);
  u8_const!(TUID, 26);
  u8_const!(OUTMRK, 27);
  u8_const!(TTYLOC, 28);
  u8_const!(_3270REGIME, 29);
  u8_const!(X3PAD, 30);
  u8_const!(NAWS, 31);
  u8_const!(TSPEED, 32);
  u8_const!(LFLOW, 33);
  u8_const!(LINEMODE, 34);
  u8_const!(XDISPLOC, 35);
  u8_const!(ENVIRON, 36);
  u8_const!(AUTHENTICATION, 37);
  u8_const!(ENCRYPT, 38);
  u8_const!(NEWENVIRON, 39);
  u8_const!(MSSP, 70);
  u8_const!(ZMP, 93);
  u8_const!(EXOPL, 255);
  u8_const!(MCCP2, 86);
  u8_const!(MCCP3, 87);
  u8_const!(GMCP, 201);
}
